--- dbt2-0.40.orig/configure.ac	2007-02-11 10:07:11.000000000 -0800
+++ dbt2-0.40/configure.ac	2010-07-19 19:02:33.000000000 -0700
@@ -459,7 +459,7 @@
   dnl ODBC Library (required)
   dnl check for `odbcinit' in -lodbc, -liodbc
 
-  if test "$use_odbc" == "yes" -a -z "$NONSP"; then 
+  if test "$use_odbc" == "yes" -a "$NONSP" == "yes" ; then 
     AC_MSG_ERROR(["It is possible to build only nonSP version of test for ODBC mode. Please 
                    use --enable-nonsp option"])
   fi
@@ -497,7 +497,7 @@
     LIBDIR="$ac_cv_use_mysql/lib/mysql"
   fi
 
-  if test -z "$NONSP" ; then
+  if test "$NONSP" == "no" ; then
     if test $MYSQL_VER -ne 0 && test $MYSQL_VER -lt 50000 ; then
       AC_MSG_ERROR(["To build SP based test you should use includes and 
                    libraries at least from MySQL 5.0 but you use them from $MYSQL_VER.
@@ -548,11 +548,11 @@
 AC_SUBST(DBLDFLAGS)
 
 dnl Check for test type(SP/NONSP)
-if test -n "$NONSP" && test "$database_to_use" != "pgsql"; then
+if test "$NONSP" == "yes" && test "$database_to_use" != "pgsql"; then
   DBC_FUNCTIONS="nonsp"  
 fi
 
-AM_CONDITIONAL(NONSP, test -n "$NONSP" && test "$database_to_use" != "pgsql")
+AM_CONDITIONAL(NONSP, test "$NONSP" == "yes" && test "$database_to_use" != "pgsql")
 
 dnl Checks for header files.
 AC_HEADER_STDC
