===== mysys/hash.c 1.42 vs edited =====
--- 1.42/mysys/hash.c	2005-03-07 01:27:17 -08:00
+++ edited/mysys/hash.c	2005-05-31 12:40:20 -07:00
@@ -270,7 +270,7 @@
   byte *rec_key= (byte*) hash_key(hash,pos->data,&rec_keylength,1);
   return ((length && length != rec_keylength) ||
 	  my_strnncoll(hash->charset, (uchar*) rec_key, rec_keylength,
-		       (uchar*) key, length));
+		       (uchar*) key, rec_keylength));
 }
 
 
