===== sql/mysql_priv.h 1.371 vs edited =====
--- 1.371/sql/mysql_priv.h	2005-10-28 09:38:09 -07:00
+++ edited/sql/mysql_priv.h	2005-10-28 10:32:44 -07:00
@@ -870,7 +870,7 @@
   External variables
 */
 
-extern time_t start_time;
+extern time_t start_time, flush_status_time;
 extern char *mysql_data_home,server_version[SERVER_VERSION_LENGTH],
 	    mysql_real_data_home[], *opt_mysql_tmpdir, mysql_charsets_dir[],
             def_ft_boolean_syntax[sizeof(ft_boolean_syntax)];
===== sql/mysqld.cc 1.606 vs edited =====
--- 1.606/sql/mysqld.cc	2005-10-28 09:38:10 -07:00
+++ edited/sql/mysqld.cc	2005-10-28 09:51:50 -07:00
@@ -345,7 +345,7 @@
 volatile ulong cached_thread_count= 0;
 
 double log_10[32];			/* 10 potences */
-time_t start_time;
+time_t start_time, flush_status_time;
 
 char mysql_home[FN_REFLEN], pidfile_name[FN_REFLEN], system_time_zone[30];
 char *default_tz_name;
@@ -2430,7 +2430,7 @@
   tzset();			// Set tzname
 
   max_system_variables.pseudo_thread_id= (ulong)~0;
-  start_time=time((time_t*) 0);
+  start_time= flush_status_time= time((time_t*) 0);
   if (init_thread_environment())
     return 1;
   mysql_init_variables();
@@ -5702,6 +5702,7 @@
   {"Threads_created",	       (char*) &thread_created,		SHOW_LONG_CONST},
   {"Threads_running",          (char*) &thread_running,         SHOW_INT_CONST},
   {"Uptime",                   (char*) 0,                       SHOW_STARTTIME},
+  {"Uptime_since_flush_status",(char*) 0,                       SHOW_FLUSHTIME},
   {NullS, NullS, SHOW_LONG}
 };
 
===== sql/sql_parse.cc 1.473 vs edited =====
--- 1.473/sql/sql_parse.cc	2005-10-28 09:38:10 -07:00
+++ edited/sql/sql_parse.cc	2005-10-28 10:32:17 -07:00
@@ -5242,6 +5242,7 @@
   }
   /* Reset the counters of all key caches (default and named). */
   process_key_caches(reset_key_cache_counters);
+  flush_status_time= time((time_t*) 0);
   pthread_mutex_unlock(&LOCK_status);
 }
 
===== sql/sql_show.cc 1.206 vs edited =====
--- 1.206/sql/sql_show.cc	2005-10-28 09:38:11 -07:00
+++ edited/sql/sql_show.cc	2005-10-28 10:31:04 -07:00
@@ -1925,6 +1925,10 @@
 	nr= (long) (thd->query_start() - start_time);
 	end= int10_to_str(nr, buff, 10);
         break;
+      case SHOW_FLUSHTIME:
+        nr= (long) (thd->query_start() - flush_status_time);
+        end= int10_to_str(nr, buff, 10);
+        break;
       case SHOW_QUESTION:
 	end= int10_to_str((long) thd->query_id, buff, 10);
         break;
===== sql/structs.h 1.41 vs edited =====
--- 1.41/sql/structs.h	2005-09-30 04:21:34 -07:00
+++ edited/sql/structs.h	2005-10-28 09:45:52 -07:00
@@ -165,7 +165,7 @@
 {
   SHOW_UNDEF,
   SHOW_LONG, SHOW_LONGLONG, SHOW_INT, SHOW_CHAR, SHOW_CHAR_PTR, SHOW_BOOL,
-  SHOW_MY_BOOL, SHOW_OPENTABLES, SHOW_STARTTIME, SHOW_QUESTION,
+  SHOW_MY_BOOL, SHOW_OPENTABLES, SHOW_STARTTIME, SHOW_FLUSHTIME, SHOW_QUESTION,
   SHOW_LONG_CONST, SHOW_INT_CONST, SHOW_HAVE, SHOW_SYS, SHOW_HA_ROWS,
 #ifdef HAVE_OPENSSL
   SHOW_SSL_CTX_SESS_ACCEPT, 	SHOW_SSL_CTX_SESS_ACCEPT_GOOD,
