diff -Naur mysql-5.1.14-original/client/client_priv.h mysql-5.1.14-beta/client/client_priv.h
--- mysql-5.1.14-original/client/client_priv.h	2007-02-08 23:48:24.000000000 -0800
+++ mysql-5.1.14-beta/client/client_priv.h	2007-02-08 23:52:06.000000000 -0800
@@ -56,7 +56,7 @@
   OPT_MYSQL_NUMBER_OF_QUERY,
   OPT_MYSQL_PRESERVE_SCHEMA,
   OPT_IGNORE_TABLE,OPT_INSERT_IGNORE,OPT_SHOW_WARNINGS,OPT_DROP_DATABASE,
-  OPT_TZ_UTC, OPT_AUTO_CLOSE, OPT_CREATE_SLAP_SCHEMA,
+  OPT_TZ_UTC, OPT_AUTO_CLOSE, OPT_AUTO_VERTICAL_OUTPUT, OPT_CREATE_SLAP_SCHEMA,
   OPT_MYSQL_REPLACE_INTO, OPT_BASE64_OUTPUT, OPT_SERVER_ID,
   OPT_FIX_TABLE_NAMES, OPT_FIX_DB_NAMES, OPT_SSL_VERIFY_SERVER_CERT,
   OPT_DEBUG_INFO, OPT_COLUMN_TYPES
diff -Naur mysql-5.1.14-original/client/mysql.cc mysql-5.1.14-beta/client/mysql.cc
--- mysql-5.1.14-original/client/mysql.cc	2007-02-08 23:48:24.000000000 -0800
+++ mysql-5.1.14-beta/client/mysql.cc	2007-02-08 23:52:11.000000000 -0800
@@ -138,7 +138,7 @@
                opt_xml=0,opt_nopager=1, opt_outfile=0, named_cmds= 0,
 	       tty_password= 0, opt_nobeep=0, opt_reconnect=1,
 	       default_charset_used= 0, opt_secure_auth= 0,
-               default_pager_set= 0, opt_sigint_ignore= 0,
+               default_pager_set= 0, opt_sigint_ignore= 0, auto_vertical_output= 0,
                show_warnings= 0, executing_query= 0, interrupted_query= 0;
 static my_bool column_types_flag;
 static ulong opt_max_allowed_packet, opt_net_buffer_length;
@@ -174,6 +174,7 @@
 static uint prompt_counter;
 static char delimiter[16]= DEFAULT_DELIMITER;
 static uint delimiter_length= 1;
+static int terminal_width = 80;
 
 #ifdef HAVE_SMEM
 static char *shared_memory_base_name=0;
@@ -225,6 +226,8 @@
 static char *get_arg(char *line, my_bool get_next_arg);
 static void init_username();
 static void add_int_to_prompt(int toadd);
+static int get_result_width(MYSQL_RES *res);
+static int get_field_disp_length(MYSQL_FIELD * field);
 
 /* A structure which contains information on the commands this program
    can understand. */
@@ -344,6 +347,9 @@
 static void nice_time(double sec,char *buff,bool part_second);
 static sig_handler mysql_end(int sig);
 static sig_handler handle_sigint(int sig);
+#if defined(HAVE_TERMIOS_H)
+static sig_handler window_resize(int sig);
+#endif
 
 int main(int argc,char *argv[])
 {
@@ -443,6 +449,13 @@
     signal(SIGINT, handle_sigint);              // Catch SIGINT to clean up
   signal(SIGQUIT, mysql_end);			// Catch SIGQUIT to clean up
 
+#if defined(HAVE_TERMIOS_H)
+  //Readline will call this if it installs a handler
+  signal(SIGWINCH, window_resize);
+  //call the SIGWINCH handler to get the default term width
+  window_resize(0);
+#endif
+
   /*
     Run in interactive mode like the ingres/postgres monitor
   */
@@ -579,6 +592,15 @@
   interrupted_query= 1;
 }
 
+#if defined(HAVE_TERMIOS_H)
+sig_handler window_resize(int sig)
+{
+  struct winsize window_size;
+
+  if (!ioctl(fileno(stdin), TIOCGWINSZ, &window_size))
+    terminal_width = (int) window_size.ws_col;
+}
+#endif
 
 static struct my_option my_long_options[] =
 {
@@ -593,6 +615,9 @@
   {"auto-rehash", OPT_AUTO_REHASH,
    "Enable automatic rehashing. One doesn't need to use 'rehash' to get table and field completion, but startup and reconnecting may take a longer time. Disable with --disable-auto-rehash.",
    (gptr*) &rehash, (gptr*) &rehash, 0, GET_BOOL, NO_ARG, 1, 0, 0, 0, 0, 0},
+  {"auto-vertical-output", OPT_AUTO_VERTICAL_OUTPUT,
+   "Automatically switch to vertical output mode if the result is wider than the terminal width.",
+   (gptr*) &auto_vertical_output, (gptr*) &auto_vertical_output, 0, GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0},
   {"no-auto-rehash", 'A',
    "No automatic rehashing. One has to use 'rehash' to get table and field completion. This gives a quicker start of mysql and disables rehashing on reconnect. WARNING: options deprecated; use --disable-auto-rehash instead.",
    0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0},
@@ -2097,11 +2122,12 @@
       else
       {
 	init_pager();
+
 	if (opt_html)
 	  print_table_data_html(result);
 	else if (opt_xml)
 	  print_table_data_xml(result);
-	else if (vertical)
+  else if (vertical || (auto_vertical_output && terminal_width < get_result_width(result)))
 	  print_table_data_vertically(result);
 	else if (opt_silent && verbose <= 2 && !output_tables)
 	  print_tab_data(result);
@@ -2429,6 +2455,33 @@
   my_afree((gptr) num_flag);
 }
 
+static int get_field_disp_length(MYSQL_FIELD * field)
+{
+  uint length = column_names ? field->name_length : 0;
+
+  if (quick)
+    length=max(length,field->length);
+  else
+    length=max(length,field->max_length);
+
+  if (length < 4 && !IS_NOT_NULL(field->flags))
+    length=4;					// Room for "NULL"
+
+  return length + 1;	
+}
+
+static int get_result_width(MYSQL_RES *result)
+{
+  unsigned int len = 0;
+  MYSQL_FIELD *field;
+
+  while ((field = mysql_fetch_field(result)))
+    len += get_field_disp_length(field) + 2;
+
+  mysql_field_seek(result, 0);	
+
+  return len + 1;	
+}
 
 static void
 tee_print_sized_data(const char *data, unsigned int data_length, unsigned int total_bytes_to_send, bool right_justified)
