--- sql/sql_lex.h.orig	2006-04-25 15:02:32.000000000 -0700
+++ sql/sql_lex.h	2006-04-25 18:23:44.000000000 -0700
@@ -112,6 +112,7 @@
   SQLCOM_SHOW_PLUGINS,
   SQLCOM_CREATE_EVENT, SQLCOM_ALTER_EVENT, SQLCOM_DROP_EVENT,
   SQLCOM_SHOW_CREATE_EVENT, SQLCOM_SHOW_EVENTS,
+  SQLCOM_SHOW_PROFILE, SQLCOM_SHOW_PROFILES,
 
   /* This should be the last !!! */
 
@@ -867,6 +868,8 @@
   enum enum_var_type option_type;
   enum enum_view_create_mode create_view_mode;
   enum enum_drop_mode drop_mode;
+  uint profile_options;
+  uint profile_query_id;
   uint uint_geom_type;
   uint grant, grant_tot_col, which_columns;
   uint fk_delete_opt, fk_update_opt, fk_match_option;
